src/psyclone/domain/lfric/arg_ordering.py

```diff
@@ -0,0 +1,714 @@
+                descriptors = self._kern.fs_descriptors
+                descriptor = descriptors.get_descriptor(unique_fs)
+                if descriptor.requires_basis:
+                    self.basis(unique_fs, var_accesses=var_accesses)
+                if descriptor.requires_diff_basis:
+                    self.diff_basis(unique_fs, var_accesses=var_accesses)
+                if descriptor.requires_orientation:
+                    self.orientation(unique_fs, var_accesses=var_accesses)
+            # Fix for boundary_dofs array to the boundary condition
+            # kernel (enforce_bc_kernel) arguments
+            if self._kern.name.lower() == "enforce_bc_code" and \
+               unique_fs.orig_name.lower() == "any_space_1":
+                self.field_bcs_kernel(unique_fs, var_accesses=var_accesses)
+
+        # Add boundary dofs array to the operator boundary condition
+        # kernel (enforce_operator_bc_kernel) arguments
+        if self._kern.name.lower() == "enforce_operator_bc_code":
+            # Sanity checks - this kernel should only have a single LMA
+            # operator as argument
+            if len(self._kern.arguments.args) > 1:
+                raise GenerationError(
+                    "Kernel {0} has {1} arguments when it should only have 1 "
+                    "(an LMA operator)".format(self._kern.name,
+                                               len(self._kern.arguments.args)))
+            op_arg = self._kern.arguments.args[0]
+            if op_arg.type != "gh_operator":
+                raise GenerationError(
+                    "Expected a LMA operator from which to look-up boundary "
+                    "dofs but kernel {0} has argument {1}.".
+                    format(self._kern.name, op_arg.type))
+            if op_arg.access != AccessType.READWRITE:
+                raise GenerationError(
+                    "Kernel {0} is recognised as a kernel which applies "
+                    "boundary conditions to an operator. However its operator "
+                    "argument has access {1} rather than gh_readwrite.".
+                    format(self._kern.name, op_arg.access.api_specific_name()))
+            self.operator_bcs_kernel(op_arg.function_space_to,
+                                     var_accesses=var_accesses)
+
+        # Reference-element properties
+        if self._kern.reference_element:
+            self.ref_element_properties(var_accesses=var_accesses)
+
+        # Mesh properties
+        if self._kern.mesh:
+            self.mesh_properties(var_accesses=var_accesses)
+
+        # Provide qr arguments if required
+        if self._kern.qr_required:
+            self.quad_rule(var_accesses=var_accesses)
+
+    def cell_position(self, var_accesses=None):
+        '''Add cell position information
+
+        :raises NotImplementedError: because this is an abstract method
+
+        '''
+        raise NotImplementedError(
```

Comment:
I left it the way it was originally (and I did notice that one or two functions were declared abstract). But since you suggested it, I am more than happy to make them all proper abstract functions (which then requires to remove one test that checked all these raise statements).
Done.

Reply:
Great. If it results in removing a test then I think we're going in the right direction.